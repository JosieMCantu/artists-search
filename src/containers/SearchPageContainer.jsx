import React from 'react';
import { useState, useEffect } from 'react';
import SearchArtistList from '../components/search/SearchArtistList';
import SearchControls from '../components/search/SearchControls';
import Pagination from '../components/pagination/Pagination'; 

//Service
const getArtist = async (url) => {
  // eslint-disable-next-line max-len
  const res = await fetch(`http://musicbrainz.org/ws/2/artist?query=${url}&fmt=json&limit=25`);
  const { artists } = await res.json();

  return artists.map((item) => ({
    id: item.id,
    name: item.name,
  }));
};

const SearchPageContainer = () => {

  const [loading, setLoading] = useState();
  const [url, setUrl] = useState('');
  const [artist, setArtist] = useState([]);
  const [currentPage, setCurrentPage] = useState(1);
  const [perPage, setPerPage] = useState(10);

  const handleChange = ({ target }) => {
    setUrl(target.value);
  };

  const handleSubmit = async (event) => {
    event.preventDefault();
    setLoading(true);
    const newArtist = await getArtist(url);
    setArtist(newArtist);
    setLoading(false);
  };

  
  useEffect(() => {
    if(url.length > 0) {
      getArtist(url)
        .then(setArtist)
        .finally(() => setLoading(false));
    }
  }, [currentPage]);
  

  const setIndex = (currentPage, perPage, artist) => {
    const lastPerPage = currentPage * perPage;
    const firstPerPage = lastPerPage - perPage;
    const result = artist.slice(firstPerPage, lastPerPage);
    return result;
  };

  const paginatedArtists = setIndex(currentPage, perPage, artist);
  const lastPage = artist.length / perPage;

  const handleButtonChange = ({ target }) => {
    if (target.value === 'next') {
      setCurrentPage(currentPage + 1);
    } else {
      setCurrentPage(currentPage - 1);
    }
  };

  if(loading){
    return (<h1>Loading...</h1>);
  }
  return (
    <>
      <SearchControls onChange={handleChange} onSubmit={handleSubmit} url={url} />
      <SearchArtistList artist={paginatedArtists} />
      <Pagination 
        onClick={handleButtonChange}
        currentPage={currentPage}
        lastPage={lastPage}
      />
    </>
  );
};

export default SearchPageContainer;
