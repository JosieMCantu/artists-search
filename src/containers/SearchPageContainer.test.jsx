/**
 * @jest-environment jsdom
 */
import React from 'react';
import SearchPageContainer from '../containers/SearchPageContainer';
import {MemoryRouter} from 'react-router-dom';
import {render, screen, waitFor} from '@testing-library/react';
import userEvent from '@testing-library/user-event';


describe('SearchPageContainer tests', () => {
    it('renders the search page to the screen', async () => {
        render(<MemoryRouter><SearchPageContainer/></MemoryRouter>)

        const input = await screen.findByLabelText('Search for an artist:')
        userEvent.type(input, 'prince')

        const submitButton = await screen.findByRole('button', {name: 'button'})

        userEvent.click(submitButton)
        

        const ul = await screen.findByRole('list', {name: 'search'})

        return waitFor(() => {expect(ul).toMatchSnapshot()})
    })
})
