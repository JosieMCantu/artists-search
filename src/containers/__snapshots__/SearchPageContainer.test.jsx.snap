// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPageContainer tests renders the search page to the screen 1`] = `
<ul
  aria-label="search"
>
  <li>
    <a
      href="/Prince/070d193a-845c-479f-980e-bef15710653e"
    >
      <p>
        Name: 
        Prince
      </p>
      <p>
        Id: 
        070d193a-845c-479f-980e-bef15710653e
      </p>
    </a>
  </li>
  <li>
    <a
      href="/Prince/4d52342d-0584-43b3-86e2-e356221c2804"
    >
      <p>
        Name: 
        Prince
      </p>
      <p>
        Id: 
        4d52342d-0584-43b3-86e2-e356221c2804
      </p>
    </a>
  </li>
  <li>
    <a
      href="/Prince/cdc0fff7-54cf-4052-a283-319b648670fd"
    >
      <p>
        Name: 
        Prince
      </p>
      <p>
        Id: 
        cdc0fff7-54cf-4052-a283-319b648670fd
      </p>
    </a>
  </li>
  <li>
    <a
      href="/Prince/33cf14b6-ddab-4453-986c-6272876f4f9f"
    >
      <p>
        Name: 
        Prince
      </p>
      <p>
        Id: 
        33cf14b6-ddab-4453-986c-6272876f4f9f
      </p>
    </a>
  </li>
  <li>
    <a
      href="/Prince/7a983e1d-dda2-42cf-bf63-78cb2bd5c6f8"
    >
      <p>
        Name: 
        Prince
      </p>
      <p>
        Id: 
        7a983e1d-dda2-42cf-bf63-78cb2bd5c6f8
      </p>
    </a>
  </li>
  <li>
    <a
      href="/Prince/e43beb57-a015-4ed9-bcd6-490674518a49"
    >
      <p>
        Name: 
        Prince
      </p>
      <p>
        Id: 
        e43beb57-a015-4ed9-bcd6-490674518a49
      </p>
    </a>
  </li>
  <li>
    <a
      href="/Prince/660697b6-9332-4c9a-8a8e-a5b4ee8c54d1"
    >
      <p>
        Name: 
        Prince
      </p>
      <p>
        Id: 
        660697b6-9332-4c9a-8a8e-a5b4ee8c54d1
      </p>
    </a>
  </li>
  <li>
    <a
      href="/Prince/6b773f61-96f1-41e4-9029-4d491dd39005"
    >
      <p>
        Name: 
        Prince
      </p>
      <p>
        Id: 
        6b773f61-96f1-41e4-9029-4d491dd39005
      </p>
    </a>
  </li>
  <li>
    <a
      href="/Bobby Prince/0b27f6b8-96db-465e-a94e-5184f63e8fad"
    >
      <p>
        Name: 
        Bobby Prince
      </p>
      <p>
        Id: 
        0b27f6b8-96db-465e-a94e-5184f63e8fad
      </p>
    </a>
  </li>
  <li>
    <a
      href="/Prince Buster/a1d32e75-2391-4d94-904e-13536304dcf6"
    >
      <p>
        Name: 
        Prince Buster
      </p>
      <p>
        Id: 
        a1d32e75-2391-4d94-904e-13536304dcf6
      </p>
    </a>
  </li>
</ul>
`;
